### PR TITLE
feat: add `--sort [TYPE]` flag 

### PR DESCRIPTION
**Pull request by sanderblue**
## What does this do / why do we need it?

First off, thank you for creating an awesome tool. We're using `git-chglog` in many open source repositories at New Relic and we would like to continue doing so.

~~The PR introduces **sorting tags by version number** instead of sorting by the date of tag creation. Users can still sort by date by passing the flag `--sort-by-date`. Using `--sort-by-date` will perform the original behavior of the `git-chglog` command.~~

Per request to integrate with the latest configuration updates from #124 (nice addition @ldelossa 🙂 ), this pull request introduces **sorting tags by semver version** using the `--sort [TYPE]` flag where `TYPE` is either `date` or `semver`.  Sort by date is still default behavior.

**Previous reasoning behind the idea:**
Since we support multiple major release branches of a repository, our CHANGELOG gets mangled when we release v1.x and then release an update for v2.x. Essentially the v1.x commits get combined with the v2.x commits in the CHANGELOG and requires manual editing that can be cumbersome with every new release we create. Sorting the tags by version number fixes this issue.

Give it a test drive:
```
make build
```
This will build the binaries for supported architectures located in the `./bin` directory- e.g. `./bin/{arch}/git-chglog` (replace `{arch}` with `darwin`, etc)

```
./bin/darwin/git-chglog
```

## How this PR fixes the problem?

Sorting the tags by version number ensures `git-chglog` selects the correct previous tag to compare to the new tag.

Current hypothetical ordering for a CHANGELOG might look like this:
```
v2.1.0 <-- will contain commit messages from v1.1.1 due to how git-chglog sorts from tag timestamp
v1.1.1 <-- patch release for v1.x (this causes an issue for all v2.x release after this one)
v2.0.0
v1.1.0
v1.0.0
```
This is because the diff that's being checked is `github.com/user/repo/compare/v1.1.1...v2.1.0`, but we need this to be `github.com/user/repo/compare/v2.0.0...v2.1.0` for the latest release in the example above.

This PR introduces ordering for a CHANGELOG that looks like this:
```
v2.1.0 
v2.0.0
v1.1.1 
v1.1.0
v1.0.0
```
This ensures the correct previous tag is selected for comparison to create the CHANGELOG entries. 

## What should your reviewer look out for in this PR?

This PR migrates this project from `dep` to [Go modules](https://blog.golang.org/migrating-to-go-modules). The `vendor` directory is no longer needed.

## Check lists

* [x] Test passed
* [x] Coding style (indentation, etc)

## Which issue(s) does this PR fix?
Resolves: #74
Resolves: #112

### CODE REVIEW

**Comment by ctrombley** [comment]
👍

**Comment by ghostsquad** [comment]
yes please!

**Comment by ghostsquad** [comment]
@sanderblue can you rebase this so we can see if we can get the tests to run & pass? We have a few new maintainers/contributors to the project, so now is the time to kick this into high gear.

This is technically not backwards compatible, so maybe we should wait to merge this and add `--sort-by-version` flag in addition, and a warning that in a future version, `--sort-by-version` will be the default?

**Comment by sanderblue** [comment]
@ghostsquad good call on the flag idea. I should be able to address this relatively soon - hopefully within the next couple days.

**Comment by khos2ow** [comment]
This sounds really cool. To be frank I haven't actually looked at the changeset yet, but wanted to suggest converting the new flag from `bool` to `string`, i.e. using `--sort TYPE` (defaults to date if not present). Then we will have all the flexibility to add more sort types in the future without polluting the flags count.

**Comment by sanderblue** [comment]
Hi @khos2ow, thanks for the feedback, but that kind of feels like premature optimization at this stage and that would also add additional code complexity because some form of validation would be needed to ensure a valid type string was passed which could grow over time. Plus right now there are only two ways to sort and I'm not sure if there would be any other ways to sort that make sense here other than maybe ascending or descending (which could be by date or by version as well). That being said, if there's a use case I'm missing I'm all ears (or eyes in this case) 🙂

**Comment by ghostsquad** [comment]
```
kubectl top pod POD_NAME --sort-by=cpu              # Show metrics for a given pod and sort it by 'cpu' or 'memory'
```

I don't think it's that far fetched to provide a `sort-by` which takes a field/type or something along those lines. It seems fairly common. 

Looking at other tools, it looks like `gsort` has a `--sort` flag that takes types but ALSO provides long & short _sort-specific_ flags:

https://www.gnu.org/software/coreutils/manual/html_node/sort-invocation.html

```
‘-g’
‘--general-numeric-sort’
‘--sort=general-numeric’


‘-h’
‘--human-numeric-sort’
‘--sort=human-numeric’

‘-M’
‘--month-sort’
‘--sort=month’

‘-n’
‘--numeric-sort’
‘--sort=numeric’

‘-R’
‘--random-sort’
‘--sort=random’
```

**Comment by Trim21** [comment]
Is it possible to add a config option at the same time?

**Comment by zlesnr** [comment]
I think I agree with @sanderblue that this feels like premature optimization to include a sort string right now, but perhaps others could change that after merge?  What other sorting methods would we implement other than the default behavior, or what this flag introduces?  Also, 10 files changed, 3 commits, and the contributor included tests.  This seems like a great addition to me.  If other folks want a config option, perhaps that can be included after merge also.

**Comment by clok** [comment]
Just adding my take that the current work is great and the call outs a sort type and config sound like great iterations to extend this feature.

File: cmd/git-chglog/main.go
Comment (clok) [inline cmd/git-chglog/main.go:130]: This will be needed after the rebase with current `master` and the bump to `urfave/cli/v2`

```suggestion
		&cli.BoolFlag{
```

File: Makefile
Comment (clok) [inline Makefile:25]: Can these changes be moved out to a separate PR please? While potentially useful, they are not related to the addition of the sort by feature.

**Review by clok — CHANGES_REQUESTED** [review]
Thanks for the work here! Like the sort by feature. Can you please rebase from `master` to pick up the changes to `urfave/cli/v2`. You will need to address one compatibility issue with the new CLI flag. 

Also, I left a comment concerning the changes to the `Makefile`. 

Awesome work!

File: Makefile
Comment (sanderblue) [reply Makefile:25]: Sure thing, no worries :)

File: cmd/git-chglog/main.go
Comment (sanderblue) [reply cmd/git-chglog/main.go]: Updated. Should be good to go now 😎

Comment (sanderblue) [reply cmd/git-chglog/main.go]: Thanks for the review as well!

**Comment by sanderblue** [comment]
@clok I've updated the PR per latest feedback. 

Thanks for all the additional reviews and feedback as well 🙂

**Review by clok — APPROVED** [review]
Thanks for the follow up. Dropped an in-line comment with a link to #112 that we can use to follow up with a more robust sort flag. 

Appreciate the follow through and contribution.

**Comment by clok** [comment]
@mavogel @khos2ow can you take a look at this PR? I think it's ready to go.

**Review by khos2ow — APPROVED** [review]
Thank you @sanderblue, LGTM!

I think you need to rebase one final time on `master` to be able to proceed with this.

**Review by clok — CHANGES_REQUESTED** [review]
@sanderblue we have merged #124 which provides this feature via a configuration option. This has been released in v0.12.0 as well. 

While the work in that PR overlaps with the work you have done in this feature PR, we do see it as a potentially complementary feature. We would like to add the CLI flag `--sort` that can provide the functionality of the config value as well. 

Would you like to rebase your feature branch and integrate the functionality from #124 with the CLI flag work you have done? Or would you like to have one us take on the work of introducing the `--sort` flag?

No pressure either way. We wanted to ask you first as you have put in this contribution and we appreciate your work. Thanks again!

**Comment by sanderblue** [comment]
@clok sounds good. I should be able to get this updated within the next day or so once I wrap up a couple things with work.

**Comment by clok** [comment]
@sanderblue thats great! Totally understand the delay. No worries there. Don't hesitate to reach out if you have questions or feedback. This work will resolve #112 as well. Thanks!

**Comment by sanderblue** [comment]
I've rebased and will be working to get this working soon with the latest updates.

**Comment by sanderblue** [comment]
@clok @khos2ow I have rebased and updated the code to use `--sort [TYPE]` as requested and it's ready for your review. This really slimmed down the pull request as well. Thanks for the patience 😎

File: cmd/git-chglog/config.go
Comment (clok) [inline cmd/git-chglog/config.go:311]: Nice. This was what I was looking for. 😀

**Review by clok — APPROVED** [review]
This looks great! Thanks again for keeping on this, following up on the feedback and updating after the config based PR came through. 🥇 

If you could update the README with the updated CLI help output in the "CLI Usage" section, that would be awesome. Just helps us keep the repo up-to-date. 

Leaving an "Approve" on this as I don't see it as critically blocking and I am happy to pick up that edit in the end.

File: cmd/git-chglog/config.go
Comment (sanderblue) [reply cmd/git-chglog/config.go:311]: Updated 😄

**Comment by sanderblue** [comment]
@clok updated the CLI usage section with the new `--sort` flag help output.

**Comment by clok** [comment]
This has been releases as a part of the [v0.14.0](https://github.com/git-chglog/git-chglog/releases/tag/v0.14.0) release. Thank you for your contribution!

**Pull request merged**
